deep_speech_2/train.py

```diff
@@ -0,0 +1,128 @@
+parser.add_argument(
+    "--use_sortagrad", default=False, type=bool, help="Use sortagrad or not.")
+parser.add_argument(
+    "--trainer_count", default=8, type=int, help="Trainer number.")
+args = parser.parse_args()
+
+
+def train():
+    """
+    DeepSpeech2 training.
+    """
+    # create network config
+    dict_size = audio_data_utils.get_vocabulary_size()
+    audio_data = paddle.layer.data(
+        name="audio_spectrogram",
+        height=161,
+        width=1000,
+        type=paddle.data_type.dense_vector(161000))
+    text_data = paddle.layer.data(
+        name="transcript_text",
+        type=paddle.data_type.integer_value_sequence(dict_size))
+    cost, _ = deep_speech2(
+        audio_data=audio_data,
+        text_data=text_data,
+        dict_size=dict_size,
+        num_conv_layers=args.num_conv_layers,
+        num_rnn_layers=args.num_rnn_layers,
+        rnn_size=args.rnn_layer_size)
+
+    # create parameters and optimizer
+    parameters = paddle.parameters.create(cost)
+    optimizer = paddle.optimizer.Adam(
+        learning_rate=5e-4, gradient_clipping_threshold=400)
+    trainer = paddle.trainer.SGD(
+        cost=cost, parameters=parameters, update_equation=optimizer)
+    # create data readers
+    feeding = {
+        "audio_spectrogram": 0,
+        "transcript_text": 1,
+    }
+    train_batch_reader_with_sortagrad = audio_data_utils.padding_batch_reader(
+        paddle.batch(
+            audio_data_utils.reader_creator(
+                manifest_path="./libri.manifest.train", sort_by_duration=True),
+            batch_size=args.batch_size // args.trainer),
+        padding=[-1, 1000])
+    train_batch_reader_without_sortagrad = audio_data_utils.padding_batch_reader(
+        paddle.batch(
+            audio_data_utils.reader_creator(
+                manifest_path="./libri.manifest.train",
+                sort_by_duration=False,
+                shuffle=True),
+            batch_size=args.batch_size // args.trainer),
+        padding=[-1, 1000])
+    test_batch_reader = audio_data_utils.padding_batch_reader(
+        paddle.batch(
+            audio_data_utils.reader_creator(
+                manifest_path="./libri.manifest.dev", sort_by_duration=False),
+            batch_size=args.batch_size // args.trainer),
+        padding=[-1, 1000])
+
+    # create event handler
+    def event_handler(event):
+        if isinstance(event, paddle.event.EndIteration):
+            if event.batch_id % 10 == 0:
+                print "/nPass: %d, Batch: %d, TrainCost: %f" % (
+                    event.pass_id, event.batch_id, event.cost)
+            else:
+                sys.stdout.write('.')
+                sys.stdout.flush()
+        if isinstance(event, paddle.event.EndPass):
+            result = trainer.test(reader=test_batch_reader, feeding=feeding)
+            print "Pass: %d, TestCost: %s" % (event.pass_id, result.cost)
+            with gzip.open("params.tar.gz", 'w') as f:
```

Comment:
Since overfitting rarely happens in DS2, it is not necessary to save multiple models with the pass index. Currently the latest one will be enough. Might add it in the future if necessary.